lib/kms/wrapper.js

```diff
@@ -0,0 +1,246 @@
+import config from '../Config';
+import inMemory from './in_memory/backend';
+import file from './file/backend';
+import common from './common';
+import async from 'async';
+import { errors } from 'arsenal';
+
+let client;
+let implName;
+
+if (config.backends.kms === 'mem') {
+    client = inMemory;
+    implName = 'memoryKms';
+} else if (config.backends.kms === 'file') {
+    client = file;
+    implName = 'fileKms';
+} else if (config.backends.kms === 'scality') {
+    throw new Error('KMS backend not implemented');
+} else {
+    throw new Error('KMS backend is not configured');
+}
+
+
+export default class kms {
+
+    /*
+     * cb(err, masterKeyId: string)
+     */
+    static createBucketKey(bucketName, log, cb) {
+        log.debug('creating a new bucket key');
+        client.createBucketKey(bucketName, log, (err, masterKeyId) => {
+            if (err) {
+                log.warn('error from kms', { implName, error: err });
+                return cb(err);
+            }
+            log.trace('bucket key created in kms');
+            return cb(null, masterKeyId);
+        });
+    }
+
+    /*
+     * cb(err, serverSideEncryptionInfo: object)masterKeyId: string)
+     */
+    static bucketLevelEncryption(bucketName, headers, log, cb) {
+        const sseAlgorithm = headers['x-amz-scal-server-side-encryption'];
+        const sseMasterKeyId =
+                  headers['x-amz-scal-server-side-encryption-aws-kms-key-id'];
+        /*
+         * the purpose of bucket level encryption is to exempt client
+         * from sending appropriate headers to trigger encryption.
+         * Since server side encryption with customer provided key needs
+         * the key for every access to the ciphered objects and we don't
+         * want to store it in the bucket metadata. It is not feasible
+         * in bucket level encryption.
+         */
+        if (sseAlgorithm === 'AES256' ||
+            (sseAlgorithm === 'aws:kms' && sseMasterKeyId === undefined)) {
+            this.createBucketKey(bucketName, log, (err, masterKeyId) => {
+                if (err) {
+                    cb(err);
+                    return;
+                }
+                const serverSideEncryptionInfo = {
+                    cryptoScheme: 1,
+                    algorithm: sseAlgorithm,
+                    masterKeyId,
+                    mandatory: true,
+                };
+                cb(null, serverSideEncryptionInfo);
+            });
+        } else if (sseAlgorithm === 'aws:kms') {
+            const serverSideEncryptionInfo = {
+                cryptoScheme: 1,
+                algorithm: sseAlgorithm,
+                masterKeyId: sseMasterKeyId,
+                mandatory: true,
+            };
+            cb(null, serverSideEncryptionInfo);
+        } else {
+            /*
+             * no encryption
+             */
+            cb(null, null);
+        }
+    }
+
+    static createDataKey(log) {
```

Comment:
is this supposed to mimic an async function?  It's sort of odd to have this function rather than just calling common.createDataKey() directly in createCipherBundle function below